0001-numerics.md

```diff
@@ -0,0 +1,93 @@
+
+```rust
+struct NumericTypeInfo<T: NumericSerialization> {
+    name: &'static str,
+    // if you have to hold more than 2^64 bits, then I have bad news for you.
+    // None for infinitely sized types
+    min_bits: Option<u64>,
+    min_value: Option<&'static T>,
+    // can a value of this type be signed
+    signed: bool,
+    // can there only be integral values? (only zeros after the comma)
+    integral: bool,
+}
+
+/// `Display` bound for textual representations
+trait NumericSerialization: core::fmt::Display {
+    /// yields a value of the above struct
+    /// FIXME: use associated constant once that's stable
+    fn type_info() -> &'static NumericTypeInfo<Self>;
+    /// yields the number of bits required to hold this value
+    fn num_bits(&self) -> u64;
+    /// whether the value is signed
+    fn is_signed(&self) -> bool;
+    /// whether the value is integral
+    fn is_integral(&self) -> bool;
+    /// yield an f64 representation of the value if that can be done losslessly
+    fn as_f64(&self) -> Option<f64>;
+    /// yield an u64 representation of the value if that can be done losslessly
+    fn as_u64(&self) -> Option<u64>;
+    /// yield an i64 representation of the value if that can be done losslessly
+    fn as_i64(&self) -> Option<i64>;
+    /// Iterate over the bits starting with the least significant one
+    /// This function uses a closure until we get `impl Trait` return types
+    /// FIXME: add more bounds to the `Iterator` to allow things like `.rev()`
+    fn iter_bits<F: FnOnce(&mut Iterator<Item=bool>)>(&self, f: F);
+}
+```
+
+With this information any numeric type can offer enough information for a `Serializer` to decide how to proceed. It's important to use static dispatch when passing this value (no other way anyway, it's not object safe, let's keep it this way). Static dispatch helps optimizing out the entire thing.
+
+Deserialization needs the design in reverse. The trait `NumericDeserialization` is again implemented for concrete types. 
+
+```rust
+trait NumericDeserialization {
+    /// FIXME: use associated constant once that's stable
+    fn type_info() -> &'static NumericTypeInfo<Self>;
+    fn from_str(&str) -> Self;
+    fn from_bits(&mut Iterator<Item=bool>) -> Self;
+    fn from_u64(u64) -> Self;
+    fn from_f64(f64) -> Self;
+    fn from_i64(i64) -> Self;
+}
+```
+
+The `Deserializer` trait would then have a method `fn deserialize_numeric<V>(&mut self, visitor: V) where V::Value: NumericDeserialization`.
+
+# Drawbacks
+
+More layers of indirection. Newcomers to serde that want to manually build `Serializer` and `Serialize` impls will have much more initial Code to read through.
+
+# Alternatives
+
+## Bignum
+Can represent any number, but requires heap allocations
```

Comment:
`NumericDeserialization` could simply get a `from_bignum` method and `NumericSerialization` could get a `as_bignum` method. We might even be able to start adding some support through features (but enabling that bignum support might influence the community towards the one we choose).